src/unitxt/metrics.py

```diff
@@ -1430,3 +1431,116 @@ def _compute(
             for k in self.k_list:
                 result[self.score_name(measure_name, k)] = measure_array[min(k, max_k)]
         return result
+
+
+class GroupedInstanceMetric(GlobalMetric):
```

Comment:
I am not sure that deriving from `GlobalMetric` is suitable here, since the underlying metric is computed per instance. So its a bit confusing with the global metric goal to allow computation which depends on a set of instances, but here we are doing per-instance computations.

I think a better option could be modifying [`InstanceMetric.process(..)`](https://github.com/IBM/unitxt/blob/main/src/unitxt/metrics.py#L373) to support grouping. After the loop that calculates the scores per instance ([this loop](https://github.com/IBM/unitxt/blob/main/src/unitxt/metrics.py#L377)), you could group the scores by the group key, if one is provided. This requires keeping track of the group key per score. Then, the code could loop over the reduction map and do the computation per group. In case there is no group key specified, there will be only one group with all the instances. Then the final step would be to averege over all group scores (which would be one score if no groups are defined). What do you think?